ircd-hybrid-8.2.31/doc/ircd.8
.\" @(#)ircd.8 7b10 25 Oct 2001
.\" $Id: ircd.8 9102 2020-01-01 09:58:57Z michael $
.TH IRCD 8 "22 October 2015" "ircd-hybrid"
.SH NAME
ircd \- ircd-hybrid: an advanced, lightweight Internet Relay Chat Daemon
.SH SYNOPSIS
.hy 0
.IP \fBircd\fP
[-configfile filename] [-klinefile filename] [-dlinefile filename]
[-xlinefile filename] [-resvfile filename] [-logfile filename] [-pidfile filename]
[-foreground] [-version] [-help]
.SH DESCRIPTION
.LP
\fIircd\fP is the server (daemon) program for the Internet Relay Chat
Program. The \fIircd\fP is a server in that its function is to "serve"
the client program \fIirc(1)\fP and other compatible programs with
messages and commands. All commands and user messages are passed
directly to the \fIircd\fP for processing and relaying to other IRC
servers.

This is \fIircd-hybrid\fP, the highly modified (and hopefully improved)
variant of the original \fIircd\fP program by Jarkko Oikarinen.
.SH OPTIONS
.TP
.B \-configfile filename
Specifies the ircd.conf file to be used for this \fIircd\fP. The option
is used to override the default ircd.conf given at compile time. By default,
this is "etc/ircd.conf" within the prefix you installed the \fIircd\fP in.
.TP
.B \-klinefile filename
.TP
.B \-dlinefile filename
.TP
.B \-xlinefile filename
.TP
.B \-resvfile filename
Specifies the database files to be used for D-lines (host bans), K-lines (hostmask bans),
X-lines (gecos bans) and RESVs (jupes) which by default are within the var/lib/
directory of your installation prefix as dline.db, kline.db, xline.db and resv.db.
.TP
.B \-logfile filename
Specifies an alternative logfile to be used than that specified in include/defaults.h
.TP
.B \-pidfile filename
Specifies the file used by the \fIircd\fP to store its process ID. The option is
used to override the default ircd.pid given at compile time.
.TP
.B \-foreground
Makes \fIircd\fP run in the foreground.
.TP
.B \-version
Makes \fIircd\fP print its version and exit.
.TP
.B \-help
Makes \fIircd\fP print available options and exit.
.SH USAGE
If you plan to connect your IRC server to an existing IRC network,
you will need to alter your local IRC configuration file (typically named
"ircd.conf") so that it will accept and make connections to other IRC servers.
This file contains the hostnames, IP addresses and sometimes passwords
for connections to other \fIircds\fP around the world.

The example reference.conf in the etc/ directory documents itself. Read it
carefully or you may expose a risk on your network simply by having your
\fIircd\fP running!
.LP
.SH BOOTING THE SERVER
It is sufficient to type \fIircd\fP at the command line to start
\fIircd-hybrid\fP into the background. If you wish to run \fIircd\fP in the
foreground (perhaps for debugging purposes), use the \fB-foreground\fP
parameter.

The \fIircd-hybrid\fP package in your distribution may ship with an rc-script
which handles this for you. In Debian, it is \fB/etc/init.d/ircd-hybrid\fP.
.SH EXAMPLE
# ircd -configfile /usr/share/ircd/ircd.conf -logfile /var/log/ircd.log

Starts \fIircd\fP with the config file /usr/share/ircd/ircd.conf and
with a log file /var/log/ircd.log. It implicitly forks back to the
console -- to prevent this, use \fB-foreground\fP.

Many of the above paths are hard-coded at compile time in the
\fBinclude/defaults.h\fP file, so it may be convenient to edit these to
suit your needs before you compile the \fIircd\fP.
.SH COPYRIGHT
(c) 1988,1989 University of Oulu, Computing Center, Finland,
.LP
(c) 1988,1989 Department of Information Processing Science,
University of Oulu, Finland
.LP
(c) 1988-1991 Jarkko Oikarinen
.LP
(c) 1997-2020 ircd-hybrid development team
.LP
For full COPYRIGHT, see the COPYING file within the IRC source package.
.LP
.RE
.SH BUGS
If you find a bug and you have a core file from \fIircd\fP as a result, use a
debugger like \fBgdb(1)\fP to process the core file and send a backtrace
to \fBbugs@ircd-hybrid.org\fP.
.SH AUTHORS
irc2.8 and earlier: Jarkko Oikarinen, currently jto@tolsun.oulu.fi.
.LP
ircd-hybrid: ircd-hybrid development team, bugs@ircd-hybrid.org.
.LP
Manual page written by Jeff Trim, jtrim@orion.cair.du.edu,
later modified by jto@tolsun.oulu.fi.
.LP
Updated by W. Campbell, wcampbel@botbay.net, Edward Brocklesby,
ejb@klamath.uucp.leguin.org.uk, and highly revised by Joshua Kwan
(joshk@triplehelix.org) for the Debian distribution.


ircd-hybrid-8.2.31/NEWS
-- Noteworthy changes in version 8.2.31 (2020-05-03)
o) macOS compatibility fixes
o) Removed 'spoof_notice' from auth::flags. With vhosts now this notice doesn't make
   much sense anymore
o) Fixed issue where ban masks might become malformed if set by remote clients/servers
o) Fixed issue with channel mode +c where high ascii characters can be erroneously
   detected as control characters
o) For a full list of all changes in this release, see https://git.io/JfsmV


-- Noteworthy changes in version 8.2.30 (2020-03-01)
o) "QUOTE HELP extban" didn't work because the extban help file hasn't been
   installed properly during 'make install'
o) Fixed broken libcrypto detection which caused the ircd not to work with
   openssl under certain circumstances
o) Extban $n of type 'acting' has been implemented. This extban prevents
   matching users from changing their nick while in the channel. Users
   with voice or above are not affected.
o) Channel mode +N has been changed so channel members with +v can change
   their nick name as well
o) For a full list of all changes in this release, see https://git.io/JfOyF


-- Noteworthy changes in version 8.2.29 (2020-02-19)
o) Extbans have been implemented. Currently supported extbans:

   Matching:
  
   $a:<account>   Matches users logged into a matching account.
   $c:<channel>   Matches users that are on the given channel. An additional
                  prefix of either @, %, or + can be specified to test for
                  certain channel privileges.
   $o:<class>     Matches IRC operators that have joined a class
                  matching the mask.
   $r:<realname>  Matches users with a matching realname.
   $s:<server>    Matches users that are connected to a server matching the mask.
   $u:<modes>     Matches users having the specified user modes set or not set.
   $z:<certfp>    Matches users having the given TLS certificate fingerprint.
  
   Acting:
  
   $j:<banmask>   Prevents matching users from joining the channel.
   $m:<banmask>   Blocks messages from matching users. Users with voice
                  or above are not affected.

   For more details, see help/extban.
o) Added 'channel::enable_extbans' configuration option. See doc/reference.conf
   for more information.
o) For a full list of all changes in this release, see https://git.io/JvgAL


-- Noteworthy changes in version 8.2.28 (2020-01-26)
o) Fixed issue with topics set by TBURST not being propagated properly to clients
o) Allow IRC operators to search for real hosts in "WHO"
o) Ban/exempt/invex masks are now also tested against realhosts to prevent clients
   from bypassing channel bans by activating a fakehost
o) For a full list of all changes in this release, see https://git.io/Jv37M


-- Noteworthy changes in version 8.2.27 (2020-01-22)
o) Experimental support for wolfSSL has been implemented. Minimum supported version is 4.3.0
o) The --enable-openssl, --enable-gnutls switches, and their disabling counterparts
   --disable-openssl and --disable-gnutls, have been replaced with the --with-tls switch
   which takes one of the following options: 'openssl', 'wolfssl', 'gnutls', and 'none'.
   If nothing has been specified, configure tries to autodetect in the following order:
   openssl/libressl -> gnutls -> wolfssl.
o) Fixed segfault with GnuTLS/libgmp in case there's no DH parameters file
   defined in 'serverinfo::ssl_dh_param_file' or if that file is missing
o) The connection timeout for connect{} blocks can now be configured via
   the 'connect::timeout' configuration directive
o) Minimum supported OpenSSL version is 1.1.1 now
o) Minimum supported GnuTLS version is 3.6.5 now
o) Supported TLSv1.3 cipher suites can now be configured explicitely via the
   new 'serverinfo::tls_cipher_suites' configuration directive
o) In the serverinfo {} block, the following configuration directives have been renamed:
     ssl_certificate_file -> tls_certificate_file
     ssl_dh_param_file -> tls_dh_param_file
     ssl_dh_elliptic_curve -> tls_supported_groups
     ssl_cipher_list -> tls_cipher_list
     ssl_message_digest_algorithm -> tls_message_digest_algorithm
o) In the operator {} block, the following configuration directives have been renamed:
     ssl_certificate_fingerprint -> tls_certificate_fingerprint
     ssl_connection_required -> tls_connection_required
o) In the connect {} block, the following configuration directives have been renamed:
     ssl_cipher_list -> tls_cipher_list
     ssl_certificate_fingerprint -> tls_certificate_fingerprint
o) For a full list of all changes in this release, see https://git.io/Jvq9g


-- Noteworthy changes in version 8.2.26 (2019-05-31)
o) The 'general::tkline_expire_notices' configuration directive has been
   replaced with user mode 'X'. *LINE expiration notices are sent to IRC
   operators with that mode set
o) Fixed issue with '/rehash conf' creating duplicated class entries
   instead of updating existing ones that are already in use
o) For a full list of all changes in this release, see https://git.io/fjEbq


-- Noteworthy changes in version 8.2.25 (2019-04-24)
o) The 'class::number_per_ip', 'class::max_local' and 'class::max_global'
   configuration directives have been replaced with just 'class::number_per_ip_local'
   and 'class::number_per_ip_global'. The 'class::max_local' basically was
   redundant as it had the same functionality as 'class::number_per_ip'
o) Adding RESVs with wildcards no longer requires administrator privileges
o) The 'general::ignore_bogus_ts' configuration option has been deprecated
o) TLSv1.1 and TLSv1.0 are no longer supported and have been disabled in
   the openssl and gnutls module
o) Minimum supported OpenSSL version is 1.0.1f now
o) Minimum supported GnuTLS version is 3.5.8 now
o) The 'serverinfo::vhost' and 'serverinfo:vhost6' configuration directives have
   been deprecated. If you need to bind() a specific address you can specify one
   in the connect {} block

ircd-hybrid-8.2.31/src/ircd.c
/*
 *  ircd-hybrid: an advanced, lightweight Internet Relay Chat Daemon (ircd)
 *
 *  Copyright (c) 1997-2020 ircd-hybrid development team
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301
 *  USA
 */

/*! \file ircd.c
 * \brief Starts up and runs the ircd.
 * \version $Id: ircd.c 9323 2020-03-29 10:40:17Z michael $
 */

#include "stdinc.h"
#include "user.h"
#include "list.h"
#include "ircd.h"
#include "channel.h"
#include "channel_mode.h"
#include "client.h"
#include "event.h"
#include "fdlist.h"
#include "hash.h"
#include "id.h"
#include "irc_string.h"
#include "ircd_signal.h"
#include "motd.h"
#include "conf.h"
#include "hostmask.h"
#include "parse.h"
#include "res.h"
#include "restart.h"
#include "rng_mt.h"
#include "s_bsd.h"
#include "log.h"
#include "server.h"
#include "server_capab.h"
#include "send.h"
#include "modules.h"
#include "ircd_getopt.h"
#include "conf_db.h"
#include "conf_class.h"
#include "ipcache.h"
#include "isupport.h"
#include "patchlevel.h"
#include "serno.h"
#include "extban.h"


struct SetOptions GlobalSetOptions;  /* /quote set variables */
struct Counter Count;
struct ServerState_t server_state;
struct ServerStatistics ServerStats;
struct Connection meConnection;  /* That's also part of me */
struct Client me = { .connection = &meConnection };  /* That's me */

char **myargv;
const char *logFileName = LPATH;
const char *pidFileName = PPATH;

bool dorehash;
bool doremotd;

static bool printVersion;

static struct lgetopt myopts[] =
{
  { "configfile", &ConfigGeneral.configfile,
   STRING, "File to use for ircd.conf" },
  { "klinefile",  &ConfigGeneral.klinefile,
   STRING, "File to use for kline database" },
  { "dlinefile",  &ConfigGeneral.dlinefile,
   STRING, "File to use for dline database" },
  { "xlinefile",  &ConfigGeneral.xlinefile,
   STRING, "File to use for xline database" },
  { "resvfile",   &ConfigGeneral.resvfile,
   STRING, "File to use for resv database" },
  { "logfile",    &logFileName,
   STRING, "File to use for ircd.log" },
  { "pidfile",    &pidFileName,
   STRING, "File to use for process ID" },
  { "foreground", &server_state.foreground,
   BOOLEAN, "Run in foreground (don't detach)" },
  { "version",    &printVersion,
   BOOLEAN, "Print version and exit" },
  { "help", NULL, USAGE, "Print this text" },
  { NULL, NULL, STRING, NULL },
};

static struct event event_cleanup_tklines =
{
  .name = "cleanup_tklines",
  .handler = cleanup_tklines,
  .when = CLEANUP_TKLINES_TIME
};

static struct event event_try_connections =
{
  .name = "try_connections",
  .handler = try_connections,
  .when = 5
};

static struct event event_comm_checktimeouts =
{
  .name = "comm_checktimeouts",
  .handler = comm_checktimeouts,
  .when = 1
};

static struct event event_save_all_databases =
{
  .name = "save_all_databases",
  .handler = save_all_databases,
  .when = DATABASE_UPDATE_TIMEOUT
};

struct event event_write_links_file =
{
  .name = "write_links_file",
  .handler = write_links_file,
};


static void
io_loop(void)
{
  while (true)
  {
    if (listing_client_list.head)
    {
      dlink_node *node = NULL, *node_next = NULL;
      DLINK_FOREACH_SAFE(node, node_next, listing_client_list.head)
        safe_list_channels(node->data, false);
    }

    /* Run pending events */
    event_run();

    comm_select();
    exit_aborted_clients();
    free_exited_clients();

    /* Check to see whether we have to rehash the configuration. */
    if (dorehash == true)
    {
      conf_rehash(true);
      dorehash = false;
    }

    if (doremotd == true)
    {
      motd_recache();
      sendto_realops_flags(UMODE_SERVNOTICE, L_ALL, SEND_NOTICE,
                           "Got signal SIGUSR1, reloading motd file(s)");
      doremotd = false;
    }
  }
}

/* initalialize_global_set_options()
 *
 * inputs       - none
 * output       - none
 * side effects - This sets all global set options needed
 */
static void
initialize_global_set_options(void)
{
  GlobalSetOptions.maxclients = ConfigServerInfo.default_max_clients;
  GlobalSetOptions.autoconn = true;
  GlobalSetOptions.spam_time = MIN_JOIN_LEAVE_TIME;
  GlobalSetOptions.spam_num = MAX_JOIN_LEAVE_COUNT;
  GlobalSetOptions.floodcount = ConfigGeneral.default_floodcount;
  GlobalSetOptions.floodtime = ConfigGeneral.default_floodtime;
  GlobalSetOptions.joinfloodcount = ConfigChannel.default_join_flood_count;
  GlobalSetOptions.joinfloodtime = ConfigChannel.default_join_flood_time;
}

/* write_pidfile()
 *
 * inputs       - filename+path of pid file
 * output       - NONE
 * side effects - write the pid of the ircd to filename
 */
static void
write_pidfile(const char *filename)
{
  FILE *fb;

  if ((fb = fopen(filename, "w")))
  {
    char buf[IRCD_BUFSIZE];
    unsigned int pid = (unsigned int)getpid();

    snprintf(buf, sizeof(buf), "%u\n", pid);

    if (fputs(buf, fb) == -1)
      ilog(LOG_TYPE_IRCD, "Error writing to pid file %s: %s",
           filename, strerror(errno));

    fclose(fb);
  }
  else
    ilog(LOG_TYPE_IRCD, "Error opening pid file %s: %s",
         filename, strerror(errno));
}

/* check_pidfile()
 *
 * inputs       - filename+path of pid file
 * output       - none
 * side effects - reads pid from pidfile and checks if ircd is in process
 *                list. if it is, gracefully exits
 * -kre
 */
static void
check_pidfile(const char *filename)
{
  FILE *fb;
  char buf[IRCD_BUFSIZE];

  if ((fb = fopen(filename, "r")))
  {
    if (fgets(buf, 20, fb) == NULL)
      ilog(LOG_TYPE_IRCD, "Error reading from pid file %s: %s",
           filename, strerror(errno));
    else
    {
      pid_t pid = atoi(buf);

      if (kill(pid, 0) == 0)
      {
        /* log(L_ERROR, "Server is already running"); */
        printf("ircd: daemon is already running\n");
        exit(EXIT_FAILURE);
      }
    }

    fclose(fb);
  }
  else if (errno != ENOENT)
    ilog(LOG_TYPE_IRCD, "Error opening pid file %s: %s",
         filename, strerror(errno));
}

/* setup_corefile()
 *
 * inputs       - nothing
 * output       - nothing
 * side effects - setups corefile to system limits.
 * -kre
 */
static void
setup_corefile(void)
{
  struct rlimit rlim; /* resource limits */

  /* Set corefilesize to maximum */
  if (getrlimit(RLIMIT_CORE, &rlim) == 0)
  {
    rlim.rlim_cur = rlim.rlim_max;
    setrlimit(RLIMIT_CORE, &rlim);
  }
}

static void
setup_fdlimit(void)
{
  struct rlimit rlim; /* resource limits */

  if (getrlimit(RLIMIT_NOFILE, &rlim))
  {
    fprintf(stderr, "getrlimit: couldn't get maximum number of file descriptors: %s\n",
            strerror(errno));
    exit(EXIT_FAILURE);
  }

#ifdef __APPLE__
  if (rlim.rlim_max > OPEN_MAX)
    rlim.rlim_max = OPEN_MAX;
#endif
  if (rlim.rlim_max > 0xFFFF)
    rlim.rlim_max = 0xFFFF;
  rlim.rlim_cur = rlim.rlim_max;

  if (setrlimit(RLIMIT_NOFILE, &rlim) == 0)
    hard_fdlimit = rlim.rlim_cur;
  else
  {
    fprintf(stderr, "setrlimit: couldn't set maximum number of file descriptors: %s\n",
            strerror(errno));
    exit(EXIT_FAILURE);
  }
}

/*
 * print_startup - print startup information
 */
static void
print_startup(int pid)
{
  printf("ircd: version %s(%s)\n", PATCHLEVEL, SERIALNUM);
  printf("ircd: pid %d\n", pid);
  printf("ircd: running in %s mode from %s\n", !server_state.foreground ? "background"
         : "foreground", ConfigGeneral.dpath);
}

static void
make_daemon(void)
{
  int pid = fork();

  if (pid < 0)
  {
    perror("fork");
    exit(EXIT_FAILURE);
  }
  else if (pid > 0)
  {
    print_startup(pid);
    exit(EXIT_SUCCESS);
  }

  setsid();
}

int
main(int argc, char *argv[])
{
  /* Check to see if the user is running us as root, which is a nono */
  if (geteuid() == 0)
  {
    fprintf(stderr, "ERROR: This server won't run as root/superuser\n");
    return -1;
  }

  /* Setup corefile size immediately after boot -kre */
  setup_corefile();

  setup_fdlimit();

  /* Save server boot time right away, so getrusage works correctly */
  event_time_set();

  /* It's not random, but it ought to be a little harder to guess */
  uint32_t seed = (uint32_t)(event_base->time.sec_real ^
                             (event_base->time.sec_monotonic | (getpid() << 16)));
  init_genrand(seed);

  ConfigGeneral.dpath      = DPATH;
  ConfigGeneral.spath      = SPATH;
  ConfigGeneral.mpath      = MPATH;
  ConfigGeneral.configfile = CPATH;    /* Server configuration file */
  ConfigGeneral.klinefile  = KPATH;    /* Server kline file         */
  ConfigGeneral.xlinefile  = XPATH;    /* Server xline file         */
  ConfigGeneral.dlinefile  = DLPATH;   /* dline file                */
  ConfigGeneral.resvfile   = RESVPATH; /* resv file                 */

  myargv = argv;
  umask(077);  /* umask 077: u=rwx,g=,o= */

  parseargs(&argc, &argv, myopts);

  if (printVersion == true)
  {
    printf("ircd: version %s(%s)\n", PATCHLEVEL, SERIALNUM);
    exit(EXIT_SUCCESS);
  }

  if (chdir(ConfigGeneral.dpath))
  {
    perror("chdir");
    exit(EXIT_FAILURE);
  }

  if (server_state.foreground == false)
  {
    make_daemon();
    close_standard_fds(); /* this needs to be before comm_select_init()! */
  }
  else
    print_startup(getpid());

  setup_signals();

  /* We need this to initialise the fd array before anything else */
  fdlist_init();
  log_set_file(LOG_TYPE_IRCD, 0, logFileName);

  comm_select_init();  /* This needs to be setup early ! -- adrian */
  tls_init();

  /* Check if there is pidfile and daemon already running */
  check_pidfile(pidFileName);

  isupport_init();
  ipcache_init();
  client_init();
  class_init();
  resolver_init();      /* Needs to be setup before the io loop */
  modules_init();
  conf_read_files(true);   /* cold start init conf files */
  capab_init();  /* Set up default_server_capabs */
  initialize_global_set_options();  /* Has to be called after conf_read_files() */
  channel_mode_init();
  extban_init();
  read_links_file();
  motd_init();
  user_modes_init();

  if (EmptyString(ConfigServerInfo.name))
  {
    ilog(LOG_TYPE_IRCD, "ERROR: No server name specified in serverinfo block.");
    exit(EXIT_FAILURE);
  }

  strlcpy(me.name, ConfigServerInfo.name, sizeof(me.name));

  /* serverinfo {} description must exist.  If not, error out.*/
  if (EmptyString(ConfigServerInfo.description))
  {
    ilog(LOG_TYPE_IRCD, "ERROR: No server description specified in serverinfo block.");
    exit(EXIT_FAILURE);
  }

  strlcpy(me.info, ConfigServerInfo.description, sizeof(me.info));

  if (EmptyString(ConfigServerInfo.sid))
  {
    ilog(LOG_TYPE_IRCD, "Generating server ID");
    generate_sid();
  }
  else
    strlcpy(me.id, ConfigServerInfo.sid, sizeof(me.id));

  init_uid();

  me.from = &me;
  me.servptr = &me;
  me.connection->created_real = event_base->time.sec_real;
  me.connection->created_monotonic = event_base->time.sec_monotonic;

  SetMe(&me);
  server_make(&me);

  hash_add_id(&me);
  hash_add_client(&me);

  dlinkAdd(&me, &me.node, &global_server_list);

  load_kline_database(ConfigGeneral.klinefile);
  load_dline_database(ConfigGeneral.dlinefile);
  load_xline_database(ConfigGeneral.xlinefile);
  load_resv_database(ConfigGeneral.resvfile);

  load_all_modules(true);
  load_conf_modules();
  load_core_modules(true);

  write_pidfile(pidFileName);

  event_addish(&event_cleanup_tklines, NULL);

  /* We want try_connections to be called as soon as possible now! -- adrian */
  /* No, 'cause after a restart it would cause all sorts of nick collides */
  event_addish(&event_try_connections, NULL);

  /* Setup the timeout check. I'll shift it later :)  -- adrian */
  event_add(&event_comm_checktimeouts, NULL);

  event_addish(&event_save_all_databases, NULL);

  if (ConfigServerHide.flatten_links_delay && event_write_links_file.active == false)
  {
    event_write_links_file.when = ConfigServerHide.flatten_links_delay;
    event_add(&event_write_links_file, NULL);
  }

  ilog(LOG_TYPE_IRCD, "Server ready. Running version: %s(%s)", PATCHLEVEL, SERIALNUM);
  io_loop();

  return 0;
}
